src/orion/algo/mofa/mofa.py

```diff
@@ -0,0 +1,390 @@
+        state_dict["duplicates"] = copy.deepcopy(self.duplicates)
+        state_dict["converged"] = self.converged
+        state_dict["roi_space"] = self.roi_space
+        state_dict["frozen_param_values"] = copy.deepcopy(self.frozen_param_values)
+        if hasattr(self, "current_trials_params"):
+            state_dict["current_trials_params"] = copy.deepcopy(
+                self.current_trials_params
+            )
+        return state_dict
+
+    def set_state(self, state_dict):
+        """Reset the state of the algorithm based on the given state_dict
+
+        :param state_dict: Dictionary representing state of an algorithm
+        """
+        super(MOFA, self).set_state(state_dict)
+        self.seed_rng(0)
+        self.rng.set_state(state_dict["rng_state"])
+        self.completed_trials = copy.deepcopy(state_dict["completed_trials"])
+        self.duplicates = copy.deepcopy(state_dict["duplicates"])
+        self.converged = state_dict["converged"]
+        self.roi_space = copy.deepcopy(state_dict["roi_space"])
+        self.frozen_param_values = copy.deepcopy(state_dict["frozen_param_values"])
+        if "current_trials_params" in state_dict:
+            self.current_trials_params = copy.deepcopy(
+                state_dict["current_trials_params"]
+            )
+        elif hasattr(self, "current_trials_params"):
+            self.current_trials_params = []
+
+    def suggest(self, num):
+        """Suggest a `num`ber of new sets of parameters.
+
+        Draws points from a prepared set of samples from an orthonal Latin hypercube.
+
+        Parameters
+        ----------
+        num: int, optional
+            Number of trials to suggest. The algorithm may return less than the number of trials
+            requested.
+
+        Returns
+        -------
+        list of trials or None
+            A list of trials representing values suggested by the algorithm. The algorithm may opt
+            out if it cannot make a good suggestion at the moment (it may be waiting for other
+            trials to complete), in which case it will return None.
+
+
+        Notes
+        -----
+        New parameters must be compliant with the problem's domain `orion.algo.space.Space`.
+
+        """
+        if not hasattr(self, "current_trials_params"):
+            self._initialize()
+
+        if self.n_trials is None or len(self.completed_trials) >= self.n_trials:
+            self._prepare_next_iteration()
+
+        trials = []
+        while (
+            len(trials) < num
+            and len(self.current_trials_params) > 0
+            and not self.is_done
+        ):
+            trial_params = self.current_trials_params.pop()
+            trial_params.update(self.frozen_param_values)
+            trial = dict_to_trial(trial_params, self.space)
+            if self.has_observed(trial):
+                similar_trial, _ = self._trials_info[self.get_id(trial)]
```

Comment:
Note to self: This will be impacted by #833

Reply:
In this particular case, you'd want to call `similar_trial = self.registry.get_existing(trial)`, or `self.registry[self.get_id(trial)]`.
You might want to take a quick look at the `Registry` class, which is in `orion.algo.registry`. It's essentially a container of Trials, which can be indexed using the trial IDs.
Anywhere where you might have been doing something like `for id, trial in self._trials_info.items()`, you should now be doing `for trial in self.registry`, basically.